src/relay/pass/type_infer.cc

```diff
@@ -22,607 +22,360 @@
-
-  Type Unify(const Type &t1, const Type& t2, Span sp);
-  Type Resolve(const Type &t);
-  Expr Resolve(const Expr &e);
-
-  /*! \brief Attempt to solve a single relation. */
-  void Solve(TypeRelationData& ty_rel);
-
-  /*! \brief Attempt to solve all pending relations.
-   *
-   * If the solver
-   */
-  SolverResult Solve(std::vector<TypeRelationData>& rels);
-
-  /*! \brief Check that all relations hold. */
-  bool RelationsHold(bool scope_only = false);
-
-  /*! \brief Visit a function node, extra flag controls behavior. */
-  CheckedExpr VisitFunction(const Function& f, bool generalize);
-
- private:
-  CheckedExpr VisitExpr_(const VarNode* op) override;
-  CheckedExpr VisitExpr_(const GlobalVarNode* op) override;
-  CheckedExpr VisitExpr_(const ConstantNode* op) override;
-  CheckedExpr VisitExpr_(const TupleNode* op) override;
-  CheckedExpr VisitExpr_(const ParamNode* op) override;
-  CheckedExpr VisitExpr_(const FunctionNode* op) override;
-  CheckedExpr VisitExpr_(const CallNode* op) override;
-  CheckedExpr VisitExpr_(const LetNode* op) override;
-  CheckedExpr VisitExpr_(const IfNode* op) override;
-  CheckedExpr VisitExpr_(const OpNode* op) override;
-};
-
-TypeInferencer::TypeInferencer() {
-  this->env = EnvironmentNode::make({});
-  this->unifier = TypeUnifierNode::make(UnionFindNode::make({}));
-}
-
-TypeInferencer::TypeInferencer(Environment env) : env(env) {
-  this->unifier = TypeUnifierNode::make(UnionFindNode::make({}));
-}
-
-CheckedExpr TypeInferencer::Infer(const Expr& expr) {
-  RELAY_LOG(INFO) << "TypeInferencer::Check expr=" << expr << std::endl;
-  CheckedExpr checked_expr = this->VisitExpr(expr);
-  RELAY_LOG(INFO) << "TypeInferencer::Check type=" << checked_expr.type
-                  << std::endl;
-  Type final_type = checked_expr.type;
-  RELAY_LOG(INFO) << "TypeInferencer::Check type_after_subst=" << final_type
-                  << std::endl;
-  checked_expr.expr->checked_type_ = final_type;
-  return checked_expr;
-}
-
-CheckedExpr TypeInferencer::VisitExpr_(const VarNode* op) {
-  auto var = GetRef<Var>(op);
-  return {var, this->context.Lookup(var)};
-}
-
-CheckedExpr TypeInferencer::VisitExpr_(const GlobalVarNode* op) {
-  GlobalVar var = GetRef<GlobalVar>(op);
-  Expr e = this->env->Lookup(var);
-  return {var, e->checked_type()};
-}
-
-CheckedExpr TypeInferencer::VisitExpr_(const ConstantNode* const_node) {
-  return {GetRef<Constant>(const_node), const_node->tensor_type()};
-}
-
-CheckedExpr TypeInferencer::VisitExpr_(const TupleNode* op) {
-  Tuple pl = GetRef<Tuple>(op);
-
-  std::vector<Expr> field_exprs;
-  std::vector<Type> field_types;
-  for (auto field = pl->fields.begin(); field != pl->fields.end(); field++) {
-    auto checked_field = Infer(*field);
-    field_exprs.push_back(checked_field.expr);
-    field_types.push_back(checked_field.type);
+  Type VisitExpr_(const ParamNode* op) final {
+    // directly handled by Funtion
+    LOG(FATAL) << "not reached";
```

Comment:
We should probably create a platform specific unreachable macro, we did this in Lean and its good to inform the compiler that some cases should never be reached.